File: csharp/DVSClient/DVSClient/address/datasets/GetDatasetsResultFuture.cs
using DVSClient.Exceptions;
using DVSClient.Server.Address.Datasets;

namespace DVSClient.Address.Datasets
{
    public class GetDatasetsResultFuture
    {
        private readonly Task<RestApiGetDatasetsResponse> _apiResponse;

        public GetDatasetsResultFuture(Task<RestApiGetDatasetsResponse> apiResponse)
        {
            _apiResponse = apiResponse;
        }

        public async Task<GetDatasetsResult> GetAsync()
        {
            try
            {
                var response = await _apiResponse;
                if (response.Error != null)
                {
                    throw EDVSException.Using(response.Error);
                }
                return new GetDatasetsResult(response);
            }
            catch (Exception e) when (e is TaskCanceledException || e is OperationCanceledException)
            {
                throw new RestApiInterruptionOrExecutionException(e);
            }
        }

        public async Task<GetDatasetsResult> GetAsync(TimeSpan timeout)
        {
            try
            {
                var response = await Task.WhenAny(_apiResponse, Task.Delay(timeout)) == _apiResponse
                    ? await _apiResponse
                    : throw new TimeoutException();

                if (response.Error != null)
                {
                    throw EDVSException.Using(response.Error);
                }
                return new GetDatasetsResult(response);
            }
            catch (Exception e) when (e is TaskCanceledException || e is OperationCanceledException || e is TimeoutException)
            {
                throw new RestApiInterruptionOrExecutionException(e);
            }
        }

        public bool Cancel(bool mayInterruptIfRunning)
        {
            return _apiResponse.IsCanceled;
        }

        public bool IsCancelled()
        {
            return _apiResponse.IsCanceled;
        }

        public bool IsDone()
        {
            return _apiResponse.IsCompleted;
        }
    }
}

File: java/src/main/java/com/experian/dvs/client/address/datasets/GetDatasetsResultFuture.java
package com.experian.dvs.client.address.datasets;

import com.experian.dvs.client.exceptions.EDVSException;
import com.experian.dvs.client.server.address.dataset.RestApiGetDatasetsResponse;

import java.util.concurrent.ExecutionException;
import java.util.concurrent.Future;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.TimeoutException;

public class GetDatasetsResultFuture implements Future<GetDatasetsResult> {

    private final Future<RestApiGetDatasetsResponse> apiResponse;

    public GetDatasetsResultFuture(Future<RestApiGetDatasetsResponse> apiResponse) {
        this.apiResponse = apiResponse;
    }

    @Override
    public boolean cancel(boolean mayInterruptIfRunning) {
        return this.apiResponse.cancel(mayInterruptIfRunning);
    }

    @Override
    public boolean isCancelled() {
        return this.apiResponse.isCancelled();
    }

    @Override
    public boolean isDone() {
        return this.apiResponse.isDone();
    }

    @Override
    public GetDatasetsResult get() throws InterruptedException, ExecutionException {
        final RestApiGetDatasetsResponse response = this.apiResponse.get();
        if (response.getError() != null) {
            throw EDVSException.using(response.getError());
        }
        return new GetDatasetsResult(response);
    }

    @Override
    public GetDatasetsResult get(long timeout, TimeUnit unit) throws InterruptedException, ExecutionException, TimeoutException {
        final RestApiGetDatasetsResponse response = this.apiResponse.get(timeout, unit);
        if (response.getError() != null) {
            throw EDVSException.using(response.getError());
        }
        return new GetDatasetsResult(response);
    }
}
